server/rest/src/main/java/org/infinispan/rest/stream/CacheChunkedStream.java

```diff
@@ -0,0 +1,177 @@
+            if (!hasElement()) {
+               T item = subscriber.peek();
+               if (item == null) {
+                  // There is the possibility the stream is just empty and completed.
+                  // We need to process the `null` until is the end of the input.
+                  if (!subscriber.isCompleted() || isEndOfInput()) {
+                     break;
+                  }
+               }
+
+               setCurrent(item);
+            }
+
+            byte b;
+            while ((b = read()) != -1) {
+               buffer.writeByte(b);
+               i++;
+            }
+            subscriber.consume();
+         }
+         if (i == 0) {
+            stuttered = true;
+            return null;
+         }
+         offset += i;
+         release = false;
+         return buffer;
+      } finally {
+         if (release) {
+            buffer.release();
+         }
+      }
+   }
+
+   @Override
+   public final void setContext(ChannelHandlerContext ctx) {
+      this.ctx = ctx;
+   }
+
+   @Override
+   public long length() {
+      return -1;
+   }
+
+   @Override
+   public long progress() {
+      return offset;
+   }
+
+   private class InternalStreamSubscriber extends DefaultSubscriber<T> {
+      private static final int QUEUE_MAX_SIZE = 32;
+      private final Queue<T> queue = new ArrayDeque<>(QUEUE_MAX_SIZE);
+      private volatile boolean hasNext;
+
+      private InternalStreamSubscriber() {
+         this.hasNext = true;
+      }
+
+      @Override
+      protected void onStart() {
+         request(QUEUE_MAX_SIZE);
+      }
+
+      @Override
+      public void onNext(T item) {
+         try {
+            queue.add(item);
+         } finally {
+            // In case we stuttered retuning null, we must notify the ChunkedWriteHandler to resume reading the new value.
+            if (stuttered) {
+               if (ctx == null) {
+                  if (logger.isTraceEnabled()) logger.trace("Chunked request stuttered and can not be resumed");
+               } else {
+                  ChunkedWriteHandler handler = ctx.pipeline().get(ChunkedWriteHandler.class);
+                  if (handler != null) {
+                     handler.resumeTransfer();
+                  } else if (logger.isTraceEnabled()){
+                     logger.trace("Channel handler ChunkedWriteHandler not found to resume after stutter");
+                  }
+               }
+            }
+         }
+      }
+
+      @Override
+      public void onError(Throwable throwable) {
+         logger.errorf(throwable, "Failed reading publisher");
+         this.hasNext = false;
+         // Cancelling might return an incomplete response for the user.
+         cancel();
+      }
+
+      @Override
+      public void onComplete() {
+         this.hasNext = false;
+      }
+
+      public boolean isCompleted() {
+         return !hasNext;
+      }
+
+      public T peek() {
+         return queue.peek();
```

Comment:
This queue is not synchronized so technically the read may not see the write. The writes are fine as they are externally synchronized by reactive stream semantics though. But to be 100% we should add synchronized blocks to the calls of the queue.